cmd/token_psa.go

```diff
@@ -0,0 +1,134 @@
+// 			clientid:     "5364defc-80e6-447b-bec6-4af8d1542cae",
+// 			clientsecret: "iE0cD8bB0yJ0dS6rO3nN1hI2wU7uA5xR4gP7lD6vM0oH0nS8dN",
+// 		}, nil
+// 	case "ds":
+// 		return &typedata{
+// 			scheme:       "mymdssdk",
+// 			realm:        "driveds.com",
+// 			clientid:     "cbf74ee7-a303-4c3d-aba3-29f5994e2dfa",
+// 			clientsecret: "X6bE6yQ3tH1cG5oA6aW4fS6hK0cR0aK5yN2wE4hP8vL8oW5gU3",
+// 		}, nil
+// 	case "opel":
+// 		return &typedata{
+// 			scheme:       "mymopsdk",
+// 			realm:        "opel.com",
+// 			clientid:     "07364655-93cb-4194-8158-6b035ac2c24c",
+// 			clientsecret: "F2kK7lC5kF5qN7tM0wT8kE3cW1dP0wC5pI6vC0sQ5iP5cN8cJ8",
+// 		}, nil
+// 	case "peugeot":
+// 		return &typedata{
+// 			scheme:       "mymap",
+// 			realm:        "peugeot.com",
+// 			clientid:     "1eebc2d5-5df3-459b-a624-20abfcf82530",
+// 			clientsecret: "T5tP7iS0cO8sC0lA2iE2aR7gK6uE5rF3lJ8pC3nO1pR7tL8vU1",
+// 		}, nil
+// 	}
+// 	return nil, errors.New("unsupported vehicle type")
+// }
+
+func psaToken(vehicleConf config.Named) (*oauth2.Token, error) {
+	// td, err := getBlub(vehicleConf.Type)
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	var country string
+	prompt_country := &survey.Input{
+		Message: "Please enter your contry code:",
+	}
+	if err := survey.AskOne(prompt_country, &country, survey.WithValidator(survey.Required)); err != nil {
+		return nil, err
+	}
+
+	brand := strings.ToLower(vehicleConf.Type)
+	cv := oauth2.GenerateVerifier()
+	oc := psa.Oauth2Config(brand, strings.ToLower(country))
+
+	// challenge := oauth2.S256ChallengeFromVerifier(cv)
+
+	// redirect_uri := fmt.Sprintf("%s://oauth2redirect/%s", td.scheme, country)
+	// authorize_url := fmt.Sprintf("https://idpcvs.%s/am/oauth2/authorize?%s", td.realm, url.Values{
+	// 	"client_id":             []string{td.clientid},
+	// 	"redirect_uri":          []string{redirect_uri},
+	// 	"response_type":         []string{"code"},
+	// 	"scope":                 []string{"openid profile"},
+	// 	"code_challenge":        []string{challenge},
+	// 	"code_challenge_method": []string{"S256"},
+	// }.Encode())
+
+	authorize_url := oc.AuthCodeURL("", oauth2.S256ChallengeOption(cv))
+
+	fmt.Println("Please visit: ", authorize_url)
+	fmt.Println("And grab the authorization code like described here: https://github.com/flobz/psa_car_controller/discussions/779")
+
+	var code string
+	prompt_code := &survey.Input{
+		Message: "Please enter your authorization code:",
+	}
+	if err := survey.AskOne(prompt_code, &code, survey.WithValidator(survey.Required)); err != nil {
+		return nil, err
+	}
+
+	client := request.NewClient(util.NewLogger(brand))
+	client.Transport = &transport.Decorator{
+		Base: client.Transport,
+		Decorator: transport.DecorateHeaders(map[string]string{
+			"Authorization": transport.BasicAuthHeader(oc.ClientID, oc.ClientSecret),
+		}),
+	}
+
+	ctx := context.WithValue(context.Background(), oauth2.HTTPClient, client)
+	return oc.Exchange(ctx, code, oauth2.VerifierOption(cv))
```

Comment:
Den Authorization Header scheinen wir hier nicht extra angeben zu müssen, der kommt vermutlich durch die oauth2 config bereits automatisch.
Zumindest funktioniert es ohne die Zeilen 102-107 genauso.
Bzw. man könnte es noch weiter vereinfachen, so wie es aussieht funktioniert es auch mit einem einfachen `ctx := context.Background()`

Reply:
@hurzhurz wenn ich mir den aktuellen Stand so anschaue dann beschleicht mich noch ein ganz anderer Verdacht. Warum machen wir dieses ganze Token-Gespeichere überhaupt? Bei Tesla ist das notwendig weil ein altes Token unwiederbringlich verbrannt ist und nicht mehr refreshed werden kann wenn es das einmal wurde. Ist das hier auch der Fall? Falls nein wäre die Identity einfach eine ganz normalen `oauth2.TokenSource` wie in anderen Integrationen auch. Jegliches Speichern und manuelle Refreshen entfiele. 

Dann kann aber nur funktionieren wenn es wirklich aufh OAuth Tokens sind. Du schriebst oben:

> PSA nutzt als Tokens UUIDs (Version 4), die komplett zufällig generiert sind...

Wie wird denn dann überhaupt ein Token Refresh ausgeführt? Woher wissen wir denn, wann das notwendig ist?

Reply:
Das ist tatsächlich auch bei PSA so, der refresh_token ist nach einmaliger Verwendung ungültig. Und der access_token läuft nach einer Stunde ab.

Beim Refresh bekommt man neben beiden Tokens auch eine expiry time für den access_token, sieht dann z.B. so aus:
{"access_token":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx","token_type":"Bearer","refresh_token":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx","expiry":"2024-04-28T22:08:25.464024098Z"}

Beim Token aus der config Datei ist es so gelöst, dass der einfach sofort als abgelaufen angesehen wird. Das habe ich aus der Mercedes Integration so übernommen.
Entsprechend wird dann sofort ein Refresh ausgeführt und der dann neue Token mit bekanntem Expiry in der DB gespeichert. Also falls alternativ in der DB nicht schon ein aktuellerer Token ist.
Das könnte man aber noch optimieren, also wie vorgeschlagen (https://github.com/evcc-io/evcc/pull/13612#issuecomment-2081522223) evtl. direkt von der token_psa.go aus in die DB zu schreiben und gar nicht manuell in die config Datei eintragen zu lassen.

Die oauth2.TokenSource kommt jedenfalls mit dem Expiry und der automatischen Verlängerung auch klar, nur für das Speichern gibt es da wohl keine Möglichkeit das automatisch erledigen zu lassen.
Ich denke mal das war auch der Grund für die EVCC eigene oauth.RefreshTokenSource, um eine eigene RefreshToken Funktion bauen zu können in der man das macht.
Und die Funktion hier hab ich ja gestern Abend noch etwas umgebaut. Nun wird dort oauth2.TokenSource zur Verlängerung verwendet statt einem selbst gebauten HTTP Request.

Reply:
Also: es sind tatsächlich JWT Tokens, keine UUIDs! Damit ist das Subject der richtige Key für die Speicherung. Auch den Token Refresh sollten wir nur machen, wenn er notwendig ist. Der ist *immer* nötig wenn wir nur ein AccessToken aus der Config haben da dann das Expirydate fehlt. Es ist nicht notwendig wenn wir ein gültiges Token (in dem Fall ein komplettes JWT Token) aus der Settings DB bekommen.
Schau Dir bitte mal an, wie das gerade bei Tesla gelöst ist. Das ist die Referenzimplementierung und sollten wir hier identisch machen.